Courseware search

martynjames opened the pull request:
@mattdrayer, @chrisndodge, @cdodge - Here's a first look at the LMS / Studio / common changes to use search app within Studio and LMS.

Please give your feedback and opions on this. Thanks.

@dsego - Notice the move from edx-solutions to edx

martynjames commented on requirements/edx/github.txt:
```diff
@@ -21,7 +21,7 @@
 
 # Our libraries:
--e git+https://github.com/edx/XBlock.git@9c634481dfc85a17dcb3351ca232d7098a38e10e#egg=XBlock
+-e git+https://github.com/edx-solutions/XBlock.git@726847de5b2fba3c666722c73e32bac73ad6e794#egg=XBlock
```
See https://github.com/edx/XBlock/pull/286 - We'll want to update this when in place.

martynjames commented on requirements/edx/github.txt:
```diff
@@ -36,3 +36,4 @@ git+https://github.com/mitocw/django-cas.git@60a5b8e5a62e63e0d5d224a87f0b489201a
 -e git+https://github.com/edx/edx-oauth2-provider.git@0.4.0#egg=oauth2-provider
 -e git+https://github.com/edx/edx-val.git@ba00a5f2e0571e9a3f37d293a98efe4cbca850d5#egg=edx-val
 -e git+https://github.com/pmitros/RecommenderXBlock.git@b41ba8778b98da0ea680ffb8bbc59492d669df2d#egg=recommender-xblock
+-e git+https://github.com/edx-solutions/edx-search.git@021485b4b721ee49e6a5598ee14f02131b244990#egg=edx-search
```
See https://github.com/edx-solutions/edx-search/pull/1 - will update this after review and creation of edx owned edx-search

chrisndodge commented on lms/templates/courseware_search/search_error.underscore:
```diff
@@ -0,0 +1 @@
+There was an error, try searching again.
```
I18N

chrisndodge commented on lms/templates/courseware_search/search_loading.underscore:
```diff
@@ -0,0 +1 @@
+loading...
```
I18N

martynjames commented:
@cdodge, @chrisndodge, @mattdrayer - rebased with squash because rebase was very difficult otherwise. (History is in mjames/courseware_search_history for your reference if desired.)

@singingwolfboy - Steve M. suggested you'd be reviewing this, so here it is. 

NB - There is one failing bokchoy test which is failing which we are working on addressing with the file-backed mock, but apparently we don't have the path set up correctly yet. However, we are ready for feedback given the current state of affairs, and how to address it quickly. Thanks!

singingwolfboy commented on common/lib/xmodule/xmodule/tests/test_video.py:
```diff
@@ -42,6 +101,7 @@ def instantiate_descriptor(**field_data):
 
 
 class VideoModuleTest(LogicTest):
+
```
Why all the extra whitespace in this pull request?

martynjames replied:
New pylint complains about 1 vs. 2 white lines within certain places

singingwolfboy replied:
I just ran pylint locally on a test file, and I didn't see any warning about not enough whitespace. Pylint has always had rules about 1 vs 2 blank lines -- you need 2 lines between class definitions, and 1 line between method definitions. However, you don't need this extra whitespace between naming the class and writing the docstring, and it fact it makes it *harder* to read. Please remove the whitespace, and if pylint errors show up as a result, let me know.

martynjames replied:
Well... the pep8 plugin for Sublime is what adds them - sorry I was unsure as to what introduced them, but basically I had the plugin reformat the file to fix the problem, and that is what adds the extra line.

Shame because it's a very useful plugin, keeping things clean - and there does not appear to be an option to not add those lines. (I suppose the authors of the login at least think that it is somehow more pep8y to have them!). I'll remove them - maybe I can edit the plugin source...

singingwolfboy replied:
Please remove this extraneous whitespace before merging this pull request.

singingwolfboy commented on lms/static/js/search/app.js:
```diff
@@ -0,0 +1,35 @@
+
+var edx = edx || {};
```
If `edx` is undefined, won't this throw a `ReferenceError`? You're better off doing something like
```js
var edx = window.edx || {};
```

dsego replied:
Don't really know, it's done that way in all the other js files in lms.

dsego replied:
In a browser, the global scope (this) is  the window object. So, except for older IEs, those two statements should be equivalent.

singingwolfboy replied:
Oh, interesting. I just tried running that statement in Chrome, Firefox, and Safari, and it seems to work correctly even when `edx` is undefined. Today I learned something new about Javascript. :)

martynjames commented:
@singingwolfboy, @cdodge, @chrisndodge, @mattdrayer - Sorry just merged some updates in from Dino that I had meant to have ready for you before (re)tagging you; my mistake.

singingwolfboy commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -283,6 +293,41 @@ def course_rerun_handler(request, course_key_string):
+                json.dumps({"status": 'success'}),
+                content_type='application/json'
+            )
+        elif request.method == 'POST':
```
You don't need to do all these checks -- the `require_http_methods(["GET"])` decorator ensures that these cases will never happen. You can also just use the `require_GET` decorator instead, which is shorter and easier to read. See the docs: https://docs.djangoproject.com/en/1.4/topics/http/decorators/

singingwolfboy commented on common/lib/xmodule/xmodule/modulestore/mongo/base.py:
```diff
@@ -945,6 +959,92 @@ def get_course(self, course_key, depth=0, **kwargs):
+        def index_item_location(item_location, current_start_date):
+            """ add this item to the search index """
+            item = _fetch_item(item_location)
+            if item:
```
What if `item` is falsy? Currently, this function will just return `None`. If that's what you want, it would be good to do that explicitly -- saves you a level of indentation, too.
```python
item = _fetch_item(item_location)
if not item:
    return None
if item.category not in ('course', 'chapter', 'sequential', 'vertical', 'html', 'video'):
    return None
if item.start and (not current_start_date or item.start > current_start_date):
    current_start_date = item.start
# ...

singingwolfboy commented on lms/static/js/spec/main.js:
```diff
@@ -514,6 +514,34 @@
                 exports: 'Annotator',
                 deps: ['jquery']
+            },
+            'js/search/views/form': {
```
Is there a reason that you're using the Backbone shim config to declare these dependencies and such, instead of using the `define` function in the Javascript files so that require.js picks them up automatically?

singingwolfboy commented on lms/templates/courseware_search/search_loading.underscore:
```diff
@@ -0,0 +1 @@
+<%= gettext("Loading&hellip;") %>
```
[We have removed ellipses from our platform.](https://github.com/edx/edx-platform/pull/6629) Please don't reintroduce them.

singingwolfboy commented on requirements/edx/base.txt:
```diff
@@ -35,6 +35,7 @@ django-threaded-multihost==1.4-1
 django-method-override==0.1.0
 djangorestframework==2.3.14
 django==1.4.18
+elasticsearch<1.0.0
```
Why are we not using the latest version of `elasticsearch`?

martynjames replied:
Because we are not currently using the latest - trying to avoid changing too many things at once

singingwolfboy replied:
Where are we using an out-of-date version of `elasticsearch`?

martynjames replied:
It's currently used in forums, and we don't want a separate installation for this - elasticsearch environments separate information by creating different indexes, and scale by configuring the cluster.

The cluster we currently have is using the earlier version. Almost everything works with the newest client except for one specific call, so we are pinning the client as recommended by elasticsearch to be compatible with the server version. When we upgrade the server, we will want to update the client.

cpennington replied:
Please pin this to a specific version, as per the other libraries. Because this requirements file is used to install to production, we want exact versions to ensure consistency between prod and staging and any local development environments.

martynjames replied:
Will do - however, this is what is recommended by elasticsearch - let me know if you want me to change back (http://elasticsearch-py.readthedocs.org/en/master/)

singingwolfboy commented on requirements/edx/github.txt:
```diff
@@ -21,7 +21,7 @@
 
 # Our libraries:
--e git+https://github.com/edx/XBlock.git@9c634481dfc85a17dcb3351ca232d7098a38e10e#egg=XBlock
+-e git+https://github.com/edx/XBlock.git@3682847a91acac6640a330fbe797ef56ce988517#egg=XBlock
```
This update needs to be called out more clearly. Who needs to review XBlock core changes?

martynjames replied:
Not really sure how to call it out more clearly - @cdodge and @cpennington already reviewed changes

singingwolfboy replied:
Oh, I didn't realize that. Sorry.

singingwolfboy commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -283,6 +293,41 @@ def course_rerun_handler(request, course_key_string):
+        if request.method == 'GET':
+            error = reindex_course_and_check_access(course_key, request.user, depth=3)
+            if error:
+                return HttpResponse(
```
Why not use the [`JsonResponse`](https://github.com/edx/edx-platform/blob/master/common/djangoapps/util/json_request.py#L32) class? Also, if this is an error response, it must use an appropriate [HTTP status code](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes), not the default `200 OK`.

singingwolfboy commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -283,6 +293,41 @@ def course_rerun_handler(request, course_key_string):
+                    content_type='application/json'
+                )
+            return HttpResponse(
+                json.dumps({"status": 'success'}),
```
You should not need to say `"status": "success"`. The HTTP status code should indicate success or failure of the response.

singingwolfboy commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -308,6 +317,28 @@ define(["jquery", "js/common_helpers/ajax_helpers", "js/views/utils/view_utils",
+                    reindexButton.trigger('click');
+                    AjaxHelpers.expectJsonRequest(requests, 'GET', '/course_index/5');
+                    AjaxHelpers.respondWithJson(requests, createMockIndexJSON(true));
+                    expect(reindexSpy).toHaveBeenCalled();
```
You need to test that the status code is 200.

singingwolfboy replied:
Bump

singingwolfboy commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -308,6 +317,28 @@ define(["jquery", "js/common_helpers/ajax_helpers", "js/views/utils/view_utils",
+                    AjaxHelpers.expectJsonRequest(requests, 'GET', '/course_index/5');
+                    AjaxHelpers.respondWithJson(requests, createMockIndexJSON(false));
+                    expect(reindexSpy).toHaveBeenCalled();
+                    expect(errorSpy).toHaveBeenCalledWith(createMockIndexJSON(false).status);
```
You need to test that the status code is not 200.

singingwolfboy commented on cms/static/js/views/course_outline.js:
```diff
@@ -209,6 +210,30 @@ define(["jquery", "underscore", "js/views/xblock_outline", "js/views/utils/view_
+                var target = $(event.currentTarget);
+                target.css('cursor', 'wait');
+                this.startReIndex().done(function(data) {
+                    if(data.status !== 'success') {
```
I assume that `this.startReIndex()` returns a [jQuery Deferred](http://api.jquery.com/category/deferred-object/) object. If so, `.done()` indicates that the object was resolved successfully -- you don't need to test `data.status == "success"`. If you want to handle the failure case, that's what `.fail()` handlers are for.

singingwolfboy commented on cms/static/js/views/course_outline.js:
```diff
@@ -209,6 +210,30 @@ define(["jquery", "underscore", "js/views/xblock_outline", "js/views/utils/view_
+            },
+
+            onIndexError: function(errors) {
+                alert(errors.join('\n'));
```
This seems like an awkward way of notifying errors. Why not use the system feedback templates? Here's an example of showing an error prompt: https://github.com/edx/edx-platform/blob/master/cms/static/js/factories/manage_users.js#L15

singingwolfboy commented on common/lib/xmodule/xmodule/modulestore/mixed.py:
```diff
@@ -322,6 +322,21 @@ def get_course(self, course_key, depth=0, **kwargs):
+
+        :param course_key: must be a CourseKey
+        """
+        assert(isinstance(course_key, CourseKey))
```
`assert` is a keyword, not a function, and is generally used without parentheses.
```python
assert isinstance(course_key, CourseKey)
```

singingwolfboy commented on common/lib/xmodule/xmodule/modulestore/mongo/base.py:
```diff
@@ -945,6 +959,92 @@ def get_course(self, course_key, depth=0, **kwargs):
+                        item_index.update(item.index_dictionary())
+                        item_index.update({
+                            'id': unicode(item.scope_ids.usage_id),
+                        })
```
It's silly and needlessly verbose to use `.update()` with just one key-value pair. Just do
```python
item_index['id'] = unicode(item.scope_ids.usage_id)
```

martynjames commented:
@chrisndodge @singingwolfboy - hoping to get thumbs up on this sometime Monday 1/26 if possible. We believe that we have responded to all of the feedback that we can.

Do let me know if there's anything else you feel needs to be changed before a potential merge.

chrisndodge commented:
Hi Marty, looks like you need a rebase.

martynjames commented:
Rebased for merge. There seems to have been 2 other items merged into XBlock code, but neither appears to have updated the hash in platform yet. Should I be changing the hash to include those (since they are new) or just keep on eye on if they change it in the meantime?

martynjames commented:
Looks like we have some unit tests that are not running... we are looking into it

singingwolfboy commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -283,6 +293,32 @@ def course_rerun_handler(request, course_key_string):
+        raise PermissionDenied()
+    course_key = CourseKey.from_string(course_key_string)
+    with modulestore().bulk_operations(course_key):
+        error = reindex_course_and_check_access(course_key, request.user)
```
If there's an error with the `reindex_course_and_check_access` function, I would expect it to raise an exception. Returning an error value is an anti-pattern in Python -- it means that the caller has to know to check for errors. Raising an exception means that the caller doesn't have to do anything unless it wants to explicitly handle the error. Why does `reindex_course_and_check_access` return an error value, rather than raising an exception on error?

dino-cikatic replied:
In this case error is just a list of stuff that could not be indexed. They are returned to a client, but I don't believe they should throw exception and block indexing process. These can happen on multiple stuff, broken video, broken subtitles, deleted data.. Seemed to me it would be enough to just list them on UI?

singingwolfboy commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -283,6 +293,32 @@ def course_rerun_handler(request, course_key_string):
+        return HttpResponse(
+            {},
+            status=200
+        )
```
Nitpick: this is short enough to be condensed onto one line.
```python
return HttpResponse({}, status=200)
```

singingwolfboy commented on cms/djangoapps/contentstore/views/tests/test_course_index.py:
```diff
@@ -330,3 +336,129 @@ def _assert_settings_link_present(response):
+        )
+
+        # create test file in which index for this test will live
+        with open(self.TEST_INDEX_FILENAME, "w+") as index_file:
```
Where is this index file actually used? As far as I can tell, it's created in the `setUp` and removed in the `tearDown`, but never referenced anywhere else.

martynjames replied:
Yes, it is solely used only for testing when we need a persistence layer for the bok choy tests to share information between Studio (to add items to the index) and LMS (to show results). In aws and devstack, elasticsearch will be the persistence layer, but bok choy does not run in an environment where elasticsearch is present, so we use a shared file for the persistence layer.

The mock index will persist to the file when the file exists, and will not when the file does not exist.

> On Jan 26, 2015, at 10:36 AM, David Baumgold <notifications@github.com> wrote:
> 
> In cms/djangoapps/contentstore/views/tests/test_course_index.py <https://github.com/edx/edx-platform/pull/6506#discussion_r23538176>:
> 
> > +        )
> > +        self.vertical = ItemFactory.create(
> > +            parent_location=self.sequential.location, category='vertical', display_name='Subsection 1'
> > +        )
> > +        self.video = ItemFactory.create(
> > +            parent_location=self.vertical.location, category="video", display_name="My Video"
> > +        )
> > +
> > +        self.html = ItemFactory.create(
> > +            parent_location=self.vertical.location, category="html", display_name="My HTML",
> > +            data="<div>This is my unique HTML content</div>",
> > +
> > +        )
> > +
> > +        # create test file in which index for this test will live
> > +        with open(self.TEST_INDEX_FILENAME, "w+") as index_file:
> Where is this index file actually used? As far as I can tell, it's created in the setUp and removed in the tearDown, but never referenced anywhere else.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/edx/edx-platform/pull/6506/files#r23538176>.
>

singingwolfboy replied:
OK. Why are you storing a pickled dict in this file? Is there a reason it has to be pickle? Could it be JSON, YAML, or some other serialization format instead? Pickle tends to bite you in the ass when you're least expecting it.

martynjames replied:
During development Json seemed to introduce some odd handling with datetime values, but I think we can solve them using a custom encoder - to be honest, I didn't look deeply into the problem because pickle seemed to work. I can change it to Json.

chrisndodge replied:
when I have to serialize datetime to JSON, I use datetime.isoformat() to convert to ISO8601 format which works well for my uses.

singingwolfboy replied:
JSON doesn't have a native serialization format for DateTime objects, but everyone just uses [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) format. I believe Django actually has a custom JSON encoder built in that can handle DateTime objects for you. See the [Django documentation about serialization](https://docs.djangoproject.com/en/1.4/topics/serialization/), or just import it yourself:
```python
import json
from django.core.serializers.json import DjangoJSONEncoder
result = json.dumps(obj, cls=DjangoJSONEncoder)
```
See the [json module docs](https://docs.python.org/2/library/json.html#json.dump).

martynjames replied:
awesome - thank you!

singingwolfboy commented on cms/djangoapps/contentstore/views/tests/test_course_index.py:
```diff
@@ -330,3 +336,129 @@ def _assert_settings_link_present(response):
+        self.assertEqual(response.status_code, 200)
+
+        response = self.client.post(index_url, {}, HTTP_ACCEPT='application/json')
+        self.assertTrue(response.content is '')
```
```python
self.assertEqual(response.content, "")
```

singingwolfboy commented on cms/envs/aws.py:
```diff
@@ -312,3 +312,7 @@
+
+if FEATURES['ENABLE_COURSEWARE_INDEX']:
+    # Use ElasticSearch for the search engine
+    SEARCH_ENGINE = "search.elastic.ElasticSearchEngine"
```
Is there a reason why this is in an `if` block? Is there any harm to having this variable set even if its unused? It would make things simpler.

martynjames replied:
If SEARCH_ENGINE is undefined or None, then indexing is skipped. This allows tests to run as fast as possible when they do not depend upon adding information to the index.

singingwolfboy commented on cms/envs/bok_choy.py:
```diff
@@ -77,6 +77,15 @@
+SEARCH_ENGINE = "search.tests.mock_search_engine.MockSearchEngine"
+# Path at which to store the mock index - can remove disable=invalid-name
+# once global settings are set to disable this error and
+# disable=no-value-for-parameter follow the pattern above
```
Why not make this global settings change in the pull request?

martynjames replied:
Do you mean in common.py? I felt that it would be better to have it set to None by default, so that indexing operations do not get in the way of other tests. In particular, we found that when we were actively backing the mock index by file that the file was growing very large, which was slowing down all tests significantly. It is possible that I was too aggressive backing it out of everything else, and we did make the adjustment to only do the file-backed index when a test required it, but it did seem safest to test search only when we desired it.

singingwolfboy replied:
You say "can remove disable=invalid-name once global settings are set to disable this error and disable=no-value-for-parameter follow the pattern above". I assumed you were talking about adding these disables to the global `pylintrc` file, which I don't think is a good idea, but it's better than saying "we can make this change when that happens". (If it never happens, these comments will just stick around in this file forever, and we'll very quickly forget the context for them.)

Or are you referring to something else? If so, you probably want to re-word this comment to make it clearer.

martynjames replied:
Ah yes, disable=invalid-name has been added to the top of this file, so this comment is out of date - unfortunately, no-value-for-parameter is very much the same - all the other settings that get their value from calls using the / path operator also have this problem.

singingwolfboy replied:
If the comment is out of date, then please fix or remove the comment.

singingwolfboy commented on cms/static/js/views/course_outline.js:
```diff
@@ -209,6 +212,33 @@ define(["jquery", "underscore", "js/views/xblock_outline", "js/views/utils/view_
+                var msg = new AlertView.Announcement({
+                        title: gettext('Course Index'),
+                        message: gettext('Course has been successfully reindexed.'),
+                        intent: 'announcement'
```
Nitpick: if you're specifying `AlertView.Announcement`, then you don't need to also specify `intent: 'annoucement'` here.

singingwolfboy commented on cms/templates/course_outline.html:
```diff
@@ -69,6 +69,11 @@ <h3 class="sr">${_("Page Actions")}</h3>
                     </a>
                 </li>
                 <li class="nav-item">
+                    <a href="#" class="button re-index" data-category="reindex" title="${_('Click to reindex current course')}">
```
I think that saying "Click to..." is unnecessary. You can just say `title="${_('Reindex current course')}"`

singingwolfboy commented on common/lib/xmodule/xmodule/html_module.py:
```diff
@@ -253,6 +254,24 @@ def non_editable_metadata_fields(self):
+        xblock_body = super(HtmlDescriptor, self).index_dictionary()
+        html_content = re.sub(r"(\s|&nbsp;|//)+", " ", html_to_text(self.data))
+        html_content = re.sub(r"<!\[CDATA\[.*\]\]>", "", html_content)
+        html_content = re.sub(r"<!--.*-->", "", html_content)
```
Can you add some comments to these lines explaining what they are doing and why? Not everyone will immediately recognize what these strings mean. Could be as simple as `# Removing HTML-encoded non-breaking space characters`, `# removing HTML CDATA`, and `# removing HTML comments`. Also, why are you specifically removing these strings, and no others? What's special about these three types of content?

singingwolfboy commented on common/lib/xmodule/xmodule/html_module.py:
```diff
@@ -253,6 +254,24 @@ def non_editable_metadata_fields(self):
+        if "content" in xblock_body:
+            xblock_body["content"].update(html_body)
+        else:
+            xblock_body.update({"content": html_body})
```
Don't do `dict.update()` with only one key-value pair -- it's silly and over-engineered. You can rewrite this line as:
```python
xblock_body["content"] = html_body
```

singingwolfboy commented on common/lib/xmodule/xmodule/html_module.py:
```diff
@@ -253,6 +254,24 @@ def non_editable_metadata_fields(self):
+            xblock_body.update({"content": html_body})
+        xblock_body.update({
+            "content_type": "HTML Content"
+        })
```
Same comment here -- don't use `dict.update()` with only one key-value pair.

singingwolfboy commented on common/lib/xmodule/xmodule/modulestore/courseware_index.py:
```diff
@@ -0,0 +1,117 @@
+        if not searcher:
+            return
+
+        course_key = location if isinstance(location, CourseLocator) else location.course_key
```
Is there a reason you have this all on one line? I feel like it would be more readable to break it up:
```python
if isinstance(location, CourseLocator):
    course_key = location
else:
    course_key = location.course_key
```

singingwolfboy commented on common/lib/xmodule/xmodule/modulestore/courseware_index.py:
```diff
@@ -0,0 +1,117 @@
+                except Exception as err:  # pylint: disable=broad-except
+                    # broad exception so that index operation does not fail on one item of many
+                    log.warning('Could not index item: %s - %s', item_location, unicode(err))
+                    error.append('Could not index item: {} <br>'.format(item_location))
```
Why the HTML `<br>` in here? It would be better to add HTML in templates.

singingwolfboy commented on common/lib/xmodule/xmodule/modulestore/mongo/base.py:
```diff
@@ -1804,3 +1810,53 @@ def ensure_indexes(self):
+    def convert_to_draft(self, location, user_id):
+        raise NotImplementedError(
+            "{function_name} needs to be implemented in subclass".format(
+                function_name=sys._getframe().f_code.co_name
```
Whoa, what's all this magic doing? The message that you're passing to `NotImplementedError` isn't adding any value, because a `NotImplementedError` implies that the method needs to be implemented. You're better off just raising the error without a message, since you don't have anything to add.
```python
def convert_to_draft(self, location, user_id):
    raise NotImplementedError()
```

martynjames replied:
Ok - will change them.

(As an aside, all these (non) implementations were necessary because the "class" line was touched causing pre-existing errors to be detected as "differences").

singingwolfboy commented on common/lib/xmodule/xmodule/modulestore/mongo/draft.py:
```diff
@@ -541,6 +544,8 @@ def delete_item(self, location, user_id, revision=None, **kwargs):
             )
         self._delete_subtree(location, as_functions)
 
+        self.do_index(location, delete=True)
```
Why are these `do_index()` calls added here? There's no comment and no explanation for what they do, why they're here, and what operations need to call `do_index()` in the future.

singingwolfboy commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -188,3 +188,17 @@ def definition_to_xml(self, resource_fs):
+        if "content" in xblock_body:
+            xblock_body["content"].update(html_body)
+        else:
+            xblock_body.update({"content": html_body})
```
Don't call `dict.update()` with only one key-value pair.

singingwolfboy commented on common/lib/xmodule/xmodule/tests/test_video.py:
```diff
@@ -1,3 +1,4 @@
+
```
Why the blank line here? This may interfere with the UTF-8 file encoding declaration.

martynjames replied:
shouldn't be - it is now removed

singingwolfboy commented on common/lib/xmodule/xmodule/video_module/video_module.py:
```diff
@@ -612,3 +613,35 @@ def _parse_video_xml(cls, xml):
+        if "content" in xblock_body:
+            xblock_body["content"].update(video_body)
+        else:
+            xblock_body.update({"content": video_body})
```
Don't call `dict.update()` with only one key-value pair.

singingwolfboy commented on common/test/acceptance/tests/lms/test_lms_courseware_search.py:
```diff
@@ -0,0 +1,194 @@
+        """
+        # create test file in which index for this test will live
+        with open(self.TEST_INDEX_FILENAME, "w+") as index_file:
+            pickle.dump({}, index_file)
```
Please use JSON or YAML rather than pickle, if at all possible.

singingwolfboy commented on lms/envs/common.py:
```diff
@@ -1042,6 +1045,11 @@
 
+# Include JS files needed for courseware search
+courseware_js += (
+    'js/search/main.js',
```
Just add this to where `courseware_js` is defined directly above, rather than defining the variable and then immediately modifying it.

singingwolfboy commented on lms/lib/courseware_search/lms_result_processor.py:
```diff
@@ -0,0 +1,100 @@
+        Property to display the url for the given location, useful for allowing navigation
+        """
+        if "course" not in self._results_fields or "id" not in self._results_fields:
+            return None
```
Shouldn't this raise an exception, instead of returning None? I'm willing to be persuaded on this, but this seems very confusing.

martynjames replied:
Hmmm... perhaps it should ;-)

If we have any results that don't have a course association or an id then the result will still be shown, without a link in the UI.

We don't have that situation currently, if we ever desire it, we can change it back.

singingwolfboy commented on lms/lib/courseware_search/lms_result_processor.py:
```diff
@@ -0,0 +1,100 @@
+            'load',
+            self.get_item(self.get_usage_key()),
+            self.get_course_key()
+        ) is False
```
Instead of `return has_access(...) is False` it would be better to say `return not has_access(...)`.

singingwolfboy commented on lms/static/js/search/collections/SearchCollection.js:
```diff
@@ -0,0 +1,88 @@
+                },
+                type: 'POST',
+                success: function (self, xhr) {
+                    self.trigger('search');
```
Seems odd that you're completely ignoring the data that you're receiving from the server. Where does the search handler get this data?

dsego replied:
It's a handled automatically by backbone. Backbone models and collections call parse whenever data is returned by the server, in fetch, and save.

singingwolfboy commented on lms/static/js/search/views/SearchForm.js:
```diff
@@ -0,0 +1,62 @@
+        submitForm: function () {
+            this.doSearch();
+            // prevent reload
+            return false;
```
This works, but it would be clearer to have this function accept an `event` argument, and call `event.preventDefault()`, which accomplishes the same thing.

singingwolfboy commented on lms/templates/courseware_search/search_list.underscore:
```diff
@@ -0,0 +1,24 @@
+<div class="search-info">
+    <%- gettext("Searching") %> <%= courseName %>
+    <div class="search-count">
+        <span class="search-count-total"><%- totalCount %></span>  <%- ngettext("result", "results", totalCount) %>
```
Not all languages will have the number before the word "results". You should do something like this:
```js
<%
  var fmt = ngettext(
    "{total} result",
    "{total} results",
    totalCount
  );
  var totalSpan = '<span class="search-count-total">' + totalCount + '</span>';
  print(fmt.replace("{total}", totalSpan);
%>
```
That way, the translators can move around the `{total}` token in the phrase as appropriate for their language.

singingwolfboy commented on lms/urls.py:
```diff
@@ -10,7 +10,8 @@
 
-urlpatterns = ('',  # nopep8
+urlpatterns = (
+    '',  # nopep8
```
Normally, you've absolutely correct that we want to put all arguments to a function call on subsequent lines when they're too long to all put on one line. However, `urlpatterns` is a special case: because the first argument means something very different from the other arguments, the normal Django pattern is to keep the first argument on the first line with the function call. (It's structured that way in all the Django documentation.) Please revert this change back to the way it was. If pylint or pep8 is complaining, talk to me -- we can look into finding a way to silence the quality checker for this specific situation. Remember, the goal is not to blindly follow the warnings and errors that pylint/pep8 give you, the goal is to improve the readability of the code.

martynjames replied:
Yes indeed - pylint it is - with the first argument in its original location:
```
83: C0330: (bad-continuation), : Wrong continued indentation.
85: C0330: (bad-continuation), : Wrong continued indentation.
```

Where line 83 is the newly added line, and
line 85 is the closing paren

I'll revert and put in the disable for that warning.

singingwolfboy commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -308,6 +323,28 @@ define(["jquery", "js/common_helpers/ajax_helpers", "js/views/utils/view_utils",
+                    reindexButton.trigger('click');
+                    AjaxHelpers.expectJsonRequest(requests, 'GET', '/course_index/5');
+                    AjaxHelpers.respondWithJson(requests, createMockIndexJSON(false));
+                    expect(reindexSpy).toHaveBeenCalled();
```
Please test that the HTTP status code of the response is not 200.

martynjames replied:
@singingwolfboy - @dino-cikatic tells me that this is not available from here.

singingwolfboy replied:
Oh wait, you're mocking the response, not getting an actual one. Never mind, sorry.

andy-armstrong commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -283,6 +294,28 @@ def course_rerun_handler(request, course_key_string):
+    GET
+        html: return status of indexing task
+    """
+    # Only global staff (PMs) are able to rerun courses during the soft launch
```
I think this comment needs to be updated (it talks about rerunning courses).

andy-armstrong commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -283,6 +294,28 @@ def course_rerun_handler(request, course_key_string):
+@login_required
+@ensure_csrf_cookie
+@require_GET
+def course_index_handler(request, course_key_string):
```
I think this should be renamed, as we already have a ``course_index`` handler which generates the course's index page. Maybe ``course_search_index_handler``?

andy-armstrong commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -283,6 +294,28 @@ def course_rerun_handler(request, course_key_string):
 
+@login_required
+@ensure_csrf_cookie
+@require_GET
```
All new APIs should be built using Django REST Framework so I think this one should be too. It would also be good to support OAuth as we won't always want to use session authentication to kick off indexing.

See this best practices guide for details:
https://openedx.atlassian.net/wiki/display/AC/API+Thoughts

chrisndodge replied:
I'd like to defer this one as a follow-on story. The current implementations replicates existing patterns in studio, even though there's a desire to migrate to a different framework.

andy-armstrong replied:
@chrisndodge I'm okay with this so that the PR can land, but we should ensure that all future work uses the approved best practices rather than relying on the existing patterns.

singingwolfboy commented:
:+1:

andy-armstrong commented on cms/templates/course_outline.html:
```diff
@@ -69,6 +69,11 @@ <h3 class="sr">${_("Page Actions")}</h3>
                     </a>
                 </li>
                 <li class="nav-item">
+                    <a href="#" class="button re-index" data-category="reindex" title="${_('Reindex current course')}">
```
Our naming convention would be to call this ``button-reindex``, not ``re-index``.

andy-armstrong commented on cms/static/js/views/pages/course_outline.js:
```diff
@@ -24,6 +24,9 @@ define(["jquery", "underscore", "gettext", "js/views/pages/base_page", "js/views
                     self.outlineView.handleAddEvent(event);
                 });
+                this.$('.button.re-index').click(function(event) {
+                    self.outlineView.handleReIndexEvent(event);
```
Why is the logic inside the child view? That view represents the rendering of the tree itself and really has nothing to do with indexing. I would prefer to see the method in this page view.

andy-armstrong commented on cms/static/js/views/course_outline.js:
```diff
@@ -209,6 +212,32 @@ define(["jquery", "underscore", "js/views/xblock_outline", "js/views/utils/view_
                 }
+            },
+
+            handleReIndexEvent: function(event) {
```
As mentioned elsewhere, I'd move all of this logic into the CourseOutlinePage.

andy-armstrong commented on common/lib/xmodule/xmodule/modulestore/courseware_index.py:
```diff
@@ -0,0 +1,133 @@
+        Add to courseware search index from given location and its children
+        """
+        error_list = []
+        # TODO - inline for now, need to move this out to a celery task
```
This definitely needs to be moved out to a background task. If I'm reading this correctly, ``do_index`` will block the caller while the indexing takes place, and will also throw exceptions that could cause the caller to fail unexpectedly. I'd maybe rename this to something like ``add_to_search_index`` to make it clear that it isn't actually performing the indexing yet. It also should be documented that index failures will not cause exceptions to be thrown.

andy-armstrong replied:
@martynjames thanks for the explanations. I'm fine with addressing this in a follow on story, with the changes you are proposing for now. It does still look though that you are raising a ``SearchIndexingError`` at the end of this method which was my original concern. Can that just be logged too, or am I reading the code incorrectly?

martynjames replied:
I apologise, you are quite correct - we created that as a result from other feedback - it should only be raised in case of a full-course index (manually initiated), and not in the automatic processing. I shall address.

andy-armstrong commented on common/lib/xmodule/xmodule/modulestore/mongo/draft.py:
```diff
@@ -713,6 +718,10 @@ def _internal_depth_first(item_location, is_root):
             self.collection.remove({'_id': {'$in': to_be_deleted}})
+
+        # Now it's been published, add the object to the courseware search index so that it appears in search results
+        self.do_index(location)
```
This is a good example where it would be bad to block waiting on the search index, and where exceptions thrown by the indexing would upset callers of ``publish``.

chrisndodge replied:
One challenge here will be that the triggering of do_index() is deep in the bowels on lib/xmodule/modulestore which is not "Django aware". So you probably won't be able to delegate to Celery at the level of the code stack. You might need to raise a Django signal and then have a handler in the CMS level of the stack to listen on the signal and schedule a celery task.

I think Celery integration should be a quick  follow-on story.

andy-armstrong replied:
Great point @chrisndodge. It seems much cleaner to me to be firing model changed notifications from the modulestore, and then search can just be bolted on rather than being wired in directly.

andy-armstrong commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -188,3 +188,16 @@ def definition_to_xml(self, resource_fs):
         return xml_object
+
+    def index_dictionary(self):
+        xblock_body = super(SequenceDescriptor, self).index_dictionary()
```
Missing doc string

andy-armstrong commented on lms/envs/aws.py:
```diff
@@ -500,3 +500,7 @@
+
+if FEATURES.get('ENABLE_COURSEWARE_SEARCH'):
+    # Use ElasticSearch as the search engine herein
+    SEARCH_ENGINE = "search.elastic.ElasticSearchEngine"
```
@jimabramson How was Elastic search configured in Student Notes? I remember there being some issue with making it the default. Should all our search APIs use a shared configuration somehow?

jimabramson replied:
@andy-armstrong edxnotes is talking to an external data API with its own elasticsearch config behind the scenes, so not a shareable configuration for this case.

andy-armstrong commented on lms/lib/courseware_search/lms_result_processor.py:
```diff
@@ -0,0 +1,100 @@
+
+        location_description = []
+        if chapter:
+            location_description.append(get_display_name('chapter', chapter))
```
Why is this file mixing single and double quotes for constants?

andy-armstrong commented on lms/static/js/search/SearchApp.js:
```diff
@@ -0,0 +1,37 @@
+;(function (define) {
```
We name our JS files with underscores, not CamelCase.

andy-armstrong commented on lms/static/js/spec/search/search_spec.js:
```diff
@@ -0,0 +1,489 @@
+
+        it ('clears results', function () {
+            this.app.form.trigger('clear');
+            expect(this.app.results.clear).toHaveBeenCalled();
```
I don't like all these tests that verify that an internal function has been called. They don't really seem like unit tests to me. I'd rather that you not use all the spies, but instead verify that the actual method produces the desired results.

andy-armstrong commented on lms/static/js/search/views/SearchListView.js:
```diff
@@ -0,0 +1,97 @@
+;(function (define) {
```
What does this convention do? Isn't RequireJS already loaded at this point?

dsego replied:
I suspect it is something to do with RequireJS-namespace-undefine.js. Define is undefined when running search in LMS and RequireJS.define is undefined in tests.

andy-armstrong replied:
Thanks @dsego. I wonder if this isn't necessary now that https://github.com/edx/edx-platform/pull/6765 has landed (which adds RequireJS support directly into the LMS).

andy-armstrong commented on lms/static/js/spec/search/search_spec.js:
```diff
@@ -0,0 +1,489 @@
+            this.form.on('search', this.onSearch);
+        });
+
+        it('prevents default action on submit', function () {
```
I find it a bit odd to write a test for whether a particular method has been performed in the implementation. I think if there are tests that verify that the form works as desired then this test isn't necessary. Otherwise we'd need such a test on every single form throughout our code.

andy-armstrong commented on lms/static/js/spec/search/search_spec.js:
```diff
@@ -0,0 +1,489 @@
+            this.item.render();
+            expect(this.item.$el).toContainHtml(this.model.attributes.content_type);
+            expect(this.item.$el).toContainHtml(this.model.attributes.excerpt);
+            expect(this.item.$el.find('a[href="'+href+'"]')).toHaveAttr('href', href);
```
If you do a find with an href in the selector, then won't it by definition have the attr? Maybe you want to just verify that a result is returned?

andy-armstrong commented on lms/static/js/spec/search/search_spec.js:
```diff
@@ -0,0 +1,489 @@
+            expect(this.item.$el).toHaveAttr('aria-label', 'search result');
+        });
+
+        it('renders underscore template', function () {
```
IMO Jasmine titles shouldn't refer to implementation details. It is more that it renders correctly than that it renders an Underscore template.

andy-armstrong commented:
@martynjames I've finished my review pass. It is looking good, although I do have a lot of comments. Is there a sandbox where I can try this out?

martynjames commented:
Sandbox is a little out of date, and I'll update tomorrow morning, but you can take a look at:
http://courseware-search.m.sandbox.edx.org

chrisndodge commented:
sorry to say that it looks like you need another rebase

cpennington commented on common/lib/xmodule/xmodule/modulestore/courseware_index.py:
```diff
@@ -0,0 +1,133 @@
+        self.error_list = error_list
+
+
+class ModuleStoreCoursewareIndexMixin(object):
```
This doesn't need to be mixed into ModuleStore, does it? Instead, it could just be a package that the ModuleStore uses to execute indexing tasks.

andy-armstrong commented:
:+1: Thanks for all the fixes. The new unit tests look great, and all my other comments are addressed. There is a single unit test failure with Student notes, but I'll rebase your branch and hopefully a second build will resolve it. I'll also try to debug it locally.

andy-armstrong commented on lms/static/js/spec/search/search_spec.js:
```diff
@@ -0,0 +1,513 @@
+) {
+    'use strict';
+
+    describe('SearchForm', function () {
```
It would be preferable if all of these tests lived under one ``describe('Search')`` function. This would allow them all to be run as a group in the debugger.

andy-armstrong commented on lms/static/js/spec/search/search_spec.js:
```diff
@@ -0,0 +1,513 @@
+        it ('performs search', function () {
+            $('.search-field').val('search string');
+            $('.search-button').trigger('click');
+            this.server.respond();
```
This test unfortunately causes a redirect to the query page. This makes it hard to run the debug the tests in a browser as the browser changes to this URL:

    suite/search/search/search/search/query

Please change it so that the submit doesn't actually happen during the test.

martynjames replied:
@andy-armstrong - I think the behaviour you were observing was a consequence of the require.js brokenness. Certainly that was the behavior I'd see by hand when require.js was broken.

andy-armstrong replied:
@martynjames It was an interesting theory, but it is still failing for me on master. Just do:

    paver test_js_dev -s lms

dsego replied:
@andy-armstrong @martynjames Now I understand what is happening, the jasmine test suite URL changes.

andy-armstrong commented:
@martynjames While debugging the Student Notes test failure (which appears to be sporadic) I discovered a couple of issues with the Jasmine tests. These are not blockers for tonight, but it would be good to fix them as they get in the way of debugging the full suite.

cptvitamin commented on lms/static/js/fixtures/search_form.html:
```diff
@@ -0,0 +1,7 @@
+<div id="courseware-search-bar" class="search-container">
+    <form role="search-form">
+        <input type="text" class="search-field"/>
```
@mattdrayer, @chrisndodge, @cdodge, this form input has no accessible label.  Should be 
```<label for="course-search-input" class="sr">Course Search</label><input id="course-search-input" type="text" class="search-field"/>```

cptvitamin commented on lms/static/js/fixtures/search_form.html:
```diff
@@ -0,0 +1,7 @@
+<div id="courseware-search-bar" class="search-container">
+    <form role="search-form">
```
@mattdrayer, @chrisndodge, @cdodge, There is no such role as "search-form".  Must be removed.

cptvitamin commented on lms/static/js/fixtures/search_form.html:
```diff
@@ -0,0 +1,7 @@
+<div id="courseware-search-bar" class="search-container">
```
@mattdrayer, @chrisndodge, @cdodge, suggest adding a role="search" to this div

cptvitamin commented on lms/static/js/fixtures/search_form.html:
```diff
@@ -0,0 +1,7 @@
+<div id="courseware-search-bar" class="search-container">
+    <form role="search-form">
+        <input type="text" class="search-field"/>
+        <button type="submit" class="search-button" aria-label="Search">search <i class="icon-search"></i></button>
```
@mattdrayer, @chrisndodge, @cdodge, aria-label is unnecessary here since the text "search" is already accessible.  Suggest adding aria-hidden to the iconfont:
```<button type="submit" class="search-button">search <i class="icon-search" aria-hidden="true"></i></button>```

cptvitamin commented on lms/static/js/fixtures/search_form.html:
```diff
@@ -0,0 +1,7 @@
+    <form role="search-form">
+        <input type="text" class="search-field"/>
+        <button type="submit" class="search-button" aria-label="Search">search <i class="icon-search"></i></button>
+        <button type="button" class="cancel-button" aria-label="Cancel"><i class="icon-remove"></i></button>
```
@mattdrayer, @chrisndodge, @cdodge, keep the aria-label here, since there is no accessible text, but add an aria-hidden to the iconfont:
```<button type="button" class="cancel-button" aria-label="Cancel"><i class="icon-remove" aria-hidden="true"></i></button>```

e-kolpakov commented:
@martynjames quite a while have passed since last activity here, but I believe this is a correct place to start the sort of communication I'm about to start.

So, I'm working on adding search support to content libraries and have some discoveries regarding this PR and some questions.

First of all, there might be a bug here: when some XBlock is deleted from published Unit and unit *is not published*, deleted XBlock got removed from index and discarding changes does not restore it. So it probably shouldn't be deleted from index right away. On the other hand, if I understand correctly, deleting from index on item deletion was added to support removing units, and it's a valid usecase.

Now, to questions:

* I believe it would be more maintainable to have a separate index for library content. This would require directly using `search_engine.search()` instead of `api.perform_request` + creating an indexer analogous to `CoursewareSearchIndexer`. Do you think it's appropriate, or should I do something else like reusing index but having dedicated document type; or still using separate index but modifying something in edx-search (probably `api`) to have a helper function?
* If I understand right XBlocks in libraries are not accessible via modulestore after deletion as they are  persisted not in `modulestore` collection, but in `modulestore.structures`. This code, however, relies on the fact that courseware contents remain available after deletion; so the basic idea is to move code responsible for deleting from index before actual deletion from modulestore. But if modulestore failed to delete an item it would render index inconsistent. Ideally it should work in transaction, but if I'm not mistaking elastic does not support external transactions.
* Why only split_draft and split modulestores are aware of search index? Also, shouldn't this indexing stuff be external to modulestores, i.e. store would only send a signal handled somewhere else?

To be more specific: [working branch](https://github.com/open-craft/edx-platform/tree/eugeny/problem-type-filtering-via-search), [diff of my changes so far](https://github.com/edx/edx-platform/compare/master...open-craft:eugeny/problem-type-filtering-via-search?expand=1).

CC @antoviaque

martynjames commented:
Thanks Eugeny!

I think that you have indeed pointed out a bug that no-one else has identified yet - the xblock becomes sort of orphaned in that the parent gets removed, but the child does not. Thanks a lot!
> I believe it would be more maintainable to have a separate index for library content. This would require directly using search_engine.search() instead of api.perform_request + creating an indexer analogous to CoursewareSearchIndexer. Do you think it's appropriate, or should I do something else like reusing index but having dedicated document type; or still using separate index but modifying something in edx-search (probably api) to have a helper function?

Yes that is pretty much the solution that I had discussed with Braden as the best approach for your case. My understanding is that you are not wanting to search the courseware - you are searching for library items with specific properties. I like the separate index personally - unless there is going to be a reason to combine searches across library items and courseware content, in which case I would consider the different document approach.

> If I understand right XBlocks in libraries are not accessible via modulestore after deletion as they are persisted not in modulestore collection, but in modulestore.structures. This code, however, relies on the fact that courseware contents remain available after deletion; so the basic idea is to move code responsible for deleting from index before actual deletion from modulestore. But if modulestore failed to delete an item it would render index inconsistent. Ideally it should work in transaction, but if I'm not mistaking elastic does not support external transactions.
The item does not need to be present in the modulestore in order to remove it from the index, so we can probably be smart about this - the code is changing soon to handle publishing changes from a signal, which means that we can’t simply move the reindexing before the delete happens.

> Why only split_draft and split modulestores are aware of search index? Also, shouldn't this indexing stuff be external to modulestores, i.e. store would only send a signal handled somewhere else?
draft modulestore is just as aware as split modulestore - indeed, our first attempt only had it in draft. Yes, it has always been in the plan that indexing would be performed after a signal, and that it what I am currently working on. We try to build software in stages, and release an MVP when we can - we released a working solution without signals in order to have the feature in place for a specific deadline - this is considered normal “agile” software development. It is my understanding that libraries are only supported in split modulestore - perhaps that is causing some confusion.

I’ll agree that the naming is confusing:
	“draft” modulestore:
		* does have both draft and published items
			* the operations on items that are in draft mode are in the file draft.py - this is NOT a different modulestore
		* does support search index
	“split” modulestore:
		* does have both draft and published items
			* the operations on items that are in draft mode are in the file split_draft.py - this is NOT a different modulestore to split
		* does support search index
Also, BEWARE - the operations in base.py do not represent operations that are mixed into both “draft” and “split” - we originally had our code in base.py and discovered that this is not used by split!!!


Thanks for the feedback! Having said all this, I’m not sure that the right place to raise this is in an old PR; it’s a bit like closing the barn door after the horse has left - I think the right thing to do is:
* Log a defect in the JIRA system (this will help identify the bug as a bug)
* Raise these questions in a direct email and reference the logged bug.

Finally, are you asking for a review of your working branch? - normally this is done by creating a PR and tagging me within.


Marty



> On Mar 18, 2015, at 7:25 AM, Eugeny Kolpakov <notifications@github.com <mailto:notifications@github.com>> wrote:
> 
> @martynjames <https://github.com/martynjames> quite a while have passed since last activity here, but I believe this is a correct place to start the sort of communication I'm about to start.
> 
> So, I'm working on adding search support to content libraries and have some discoveries regarding this PR and some questions.
> 
> First of all, there might be a bug here: when some XBlock is deleted from published Unit and unit is not published, deleted XBlock got removed from index and discarding changes does not restore it. So it probably shouldn't be deleted from index right away. On the other hand, if I understand correctly, deleting from index on item deletion was added to support removing units, and it's a valid usecase.
> 
> Now, to questions:
> 
> I believe it would be more maintainable to have a separate index for library content. This would require directly using search_engine.search() instead of api.perform_request + creating an indexer analogous to CoursewareSearchIndexer. Do you think it's appropriate, or should I do something else like reusing index but having dedicated document type; or still using separate index but modifying something in edx-search (probably api) to have a helper function?
> If I understand right XBlocks in libraries are not accessible via modulestore after deletion as they are persisted not in modulestore collection, but in modulestore.structures. This code, however, relies on the fact that courseware contents remain available after deletion; so the basic idea is to move code responsible for deleting from index before actual deletion from modulestore. But if modulestore failed to delete an item it would render index inconsistent. Ideally it should work in transaction, but if I'm not mistaking elastic does not support external transactions.
> Why only split_draft and split modulestores are aware of search index? Also, shouldn't this indexing stuff be external to modulestores, i.e. store would only send a signal handled somewhere else?
> To be more specific: working branch <https://github.com/open-craft/edx-platform/tree/eugeny/problem-type-filtering-via-search>, diff of my changes so far <https://github.com/edx/edx-platform/compare/master...open-craft:eugeny/problem-type-filtering-via-search?expand=1>.
> 
> CC @antoviaque <https://github.com/antoviaque>
> —
> Reply to this email directly or view it on GitHub <https://github.com/edx/edx-platform/pull/6506#issuecomment-82912066>.
>

e-kolpakov commented:
@martynjames thank you for the feedback!

I agree that this is probably not the right place to discuss all of these, but I was not sure what would be the appropriate way (e.g. JIRA, personal email, etc.) So I'll create a JIRA ticket on Monday (i'm on a short vacation now) and ask all the questions I have than.

Thank you for the explanation on draft.py and split_draft.py - now it makes sense.

> The item does not need to be present in the modulestore in order to remove it from the index

Actually, it does. Code that deletes an item first fetches it from modulestore (at least it is so in current master), so index is not affected if an item already deleted from modulestore. I've spent about an hour discovering the reasons for this, so I'm pretty much sure. Alternative to removing from index before removing from modulestore would be to delete by location passed to `add_to_search_index`, but the problem is it would not delete nested XBlocks from index. I think I'll investigate it a bit further and have more observations and possible approaches to solve this next time.

> draft modulestore is just as aware as split modulestore - indeed, our first attempt only had it in draft. Yes, it has always been in the plan that indexing would be performed after a signal, and that it what I am currently working on.

Great! I might have failed to express it, but the reason I was asking this is something like "is it going to be done soon, or are there good reasons not to do that?" so I wouldn't duplicate the work or do what shouldn't be done.

>  are you asking for a review of your working branch?

No, I've included links to working branch and diff so you could have something to look at if you want to. Actually we do an internal review before or just after submitting a PR.

martynjames commented:
"The item does not need to be present in the modulestore in order to remove it from the index" - this statement _is_ true. I'll agree with you that it will not get deleted from the *current* code because of the call the fetch the item - my point is that there is no reason that this should be the case, it's just the way that we have written it; as I said, the code is changing soon, so we need to accommodate that situation anyway.
